### PR TITLE
[RFC] provider: Support for DigitalOcean

### PR DESCRIPTION
**Pull request by secant**
Still need to:
- [ ] support floating IPs
- [ ] get test coverage

### CODE REVIEW

**Comment by tejasmanohar** [comment]
We should probably use Digital Ocean's Block Storage (alike AWS EBS).

**Comment by secant** [comment]
@tejasmanohar [That is what I'm using](https://github.com/NetSys/quilt/pull/599/commits/16efcaf2dc96569332b3a17a8a4c6b127dd71336#diff-4908ea42fa76029aebdad7bf617013bbR197).

File: cluster/provider/digitalocean.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+		return nil
+	}
+	if len(bootSet) <= 0 {
+		return nil
+	}
+
+	done := make(chan error, len(bootSet))
+	errors := []error{}
+	for _, m := range bootSet {
+		go clst.createAndAttach(m, done)
```
Comment (kklin) [inline cluster/provider/digitalocean.go]: I think it makes more sense for `createAndAttach` to not know about the parallelism.

```
go func(m machine) {
    defer wg.Done()
    if err := clst.createAndAttach(m) ; err != nil {
        errChan <- err
    }
}
```

BTW, I think it'd be reasonable to just return any of the multiple errors. So the error channel would be of size 1, and we'd try a non-blocking write if there's an error.

File: cluster/provider/digitalocean.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+		return
+	}
+
+	_, _, err = clst.client.StorageActions.Attach(v.ID, d.ID)
+	if err != nil {
+		done <- err
+		return
+	}
+
+	for len(d.VolumeIDs) == 0 {
```
Comment (kklin) [inline cluster/provider/digitalocean.go]: What if this loops forever?

Comment (kklin) [inline cluster/provider/digitalocean.go]: Same for the above Status loop.

I'd write some sort of waiter function that tries every `x` seconds, and handles timing out.

File: cluster/provider/digitalocean.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+func (clst doCluster) Stop(machines []Machine) error {
+	if len(machines) <= 0 {
+		return nil
+	}
+
+	done := make(chan error, len(machines))
+	errors := []error{}
+
+	for _, m := range machines {
+		go clst.deleteAndWait(m.ID, done)
```
Comment (kklin) [inline cluster/provider/digitalocean.go]: Same comment here re parallelism.

File: cluster/provider/digitalocean.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+		return
+	}
+	d, _, _ := clst.client.Droplets.Get(id)
+	volIds := d.VolumeIDs
+	_, err = clst.client.Droplets.Delete(id)
+	if err != nil {
+		done <- err
+		return
+	}
+	for d != nil {
```
Comment (kklin) [inline cluster/provider/digitalocean.go]: Same here re waiting.

File: cluster/provider/digitalocean.go (hunk 1 above)
Comment (kklin) [inline cluster/provider/digitalocean.go]: Also easier to unit test this way.

File: cluster/provider/digitalocean.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+		}
+
+		opt.Page = page + 1
+	}
+	return machines, nil
+}
+
+// Boot will boot every machine in a goroutine, and wait for the machines to come up.
+func (clst doCluster) Boot(bootSet []Machine) error {
+	if first {
```
Comment (ejj) [inline cluster/provider/digitalocean.go]: what's this all about?  Seems weird

File: cluster/provider/digitalocean.go (hunk 2 above)
Comment (kklin) [inline cluster/provider/digitalocean.go]: BTW, @tejasmanohar added a generic waiter in his PR.

File: cluster/provider/digitalocean.go (hunk 5 above)
Comment (secant) [inline cluster/provider/digitalocean.go]: leftovers from when `List` wasn't working and it would boot many many times

**Pull request closed without merging**